Improve accessibility of React components

ivababukova opened the pull request:
Fixes #1467 and #787

#### Background

#### Change List
- Checked each Material UI component and made sure its usage conforms with its accessibility guidelines
- Added aria labels 
- Fixed this issue with colours to improve accessibility #787
- Made the dropdowns in `PopperMenu` clickable on touch screen

#### Checklist
 - [ ] Ensure PR works with all demos on the dev.vitessce.io homepage
 - [ ] Open (draft) PR's into [`vitessce-python`](https://github.com/vitessce/vitessce-python) and [`vitessce-r`](https://github.com/vitessce/vitessce-r) if this is a release PR
 - [ ] Documentation added or updated

keller-mark commented on packages/vit-s/src/shared-plot-options/CellColorEncodingOption.js:
```diff
@@ -20,12 +20,13 @@ export default function CellColorEncodingOption(props) {
-    <TableRow>
-      <TableCell className={classes.labelCell} htmlFor="cell-color-encoding-select">
+    <TableRow key="cell-color-encoding">
+      <TableCell className={classes.labelCell} variant="head" scope="row">
         {observationsLabelNice} Color Encoding
       </TableCell>
```
```suggestion
      <TableCell className={classes.labelCell} variant="head" scope="row">
        <label htmlFor="cell-color-encoding-select">
          {observationsLabelNice} Color Encoding
        </label>
      </TableCell>
```
Maybe we should keep the `htmlFor` by using a label - although we might need to adjust the styling of the label

ivababukova replied:
I have changed this to label in all places where it was referenced.

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -34,17 +34,18 @@ export default function FeatureListOptions(props) {
       {children}
-      <TableRow>
-        <TableCell className={classes.labelCell} htmlFor="feature-list-sort-option-select">
+      <TableRow key="feature-list-sort-option">
```
What is the purpose of the `key`s? I thought keys were mainly intended for rendering arrays of components https://react.dev/learn/rendering-lists#rules-of-keys

ivababukova replied:
Yeah, I did a bit more reading and you are both right. I thought that they improve accessibility, looks like I got confused. I will remove them shortly, thanks for the catch.

ivababukova replied:
Done in commit 34a756fc

keller-mark commented on packages/view-types/layer-controller/src/styles.js:
```diff
@@ -9,6 +9,16 @@ const useSpanStyles = makeStyles(() => ({
+export const channelSliderStyles = makeStyles(theme => ({
+  valueLabel: {
+    marginTop: '7px',
+    '& *': {
```
Can this selector be more specific?

ivababukova replied:
Yes, done in this commit: ee2c7380

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -121,124 +125,138 @@ export default function ScatterplotOptions(props) {
-      </TableRow>
-      <TableRow>
-        <TableCell className={classes.labelCell}>
+      { cellSetLabelsVisible && (
```
Is this change to using rendering vs. the `disabled` prop related to accessibility?

ivababukova replied:
This change has nothing to do with accessibility -- I did it for UX reasons. I thought it was confusing if the cell set labels are not visible, but the slider is still showing in a disabled mode. I thought it is confusing for the user to have a setting that is disabled. That said, you know your users better than I do, so if you disagree with this, I am happy to bring back the previous version of the code. Just let me know

ivababukova replied:
I read the comments further and I see that both of you are on the opinion that using `disabled` is better, so I reverted this change in commit e5371cbf

keller-mark commented on packages/view-types/spatial/src/SpatialOptions.js:
```diff
@@ -28,16 +28,16 @@ const ToggleFixedAxisButton = ({
-  return (
-    <TableRow>
-      <TableCell className={classes.cameraLabel}>
+  return use3d && (
```
Is this change to using rendering vs. the `disabled` prop related to accessibility?

ivababukova replied:
It is done for the same reasons as this one: https://github.com/vitessce/vitessce/pull/1612/files#r1266860907 Let me know if I should revert it.

ivababukova replied:
I read the comments further and I see that both of you are on the opinion that using `disabled` is better, so I reverted this change in commit e5371cbf

keller-mark commented on packages/view-types/spatial/src/SpatialOptions.js:
```diff
@@ -83,25 +83,25 @@ export default function SpatialOptions(props) {
   return (
     <OptionsContainer>
-      {canShowColorEncodingOption ? (
+      {canShowColorEncodingOption && (
```
Can we revert this and the other analogous changes? It does not seem related to accessibility https://github.com/vitessce/vitessce/blob/main/dev-docs/design-guidelines.md#jsx

ivababukova replied:
Done in commit 4e4f4a55

keller-mark commented:
We should also start using https://react.dev/reference/react/useId

NickAkhmetov commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -121,124 +125,138 @@ export default function ScatterplotOptions(props) {
             onChange={handleTooltipsVisibilityChange}
             name="scatterplot-option-toltip-visibility"
```
```suggestion
            name="scatterplot-option-tooltip-visibility"
```

ivababukova replied:
fixed here: c6ee12fa

NickAkhmetov reviewed (commented):
Thanks for taking on this monumental task!

I have some suggestions after reviewing `tooltip` and `view-types`; I will continue my review picking up from the `vit-s` directory tomorrow. 

To sum up my thoughts so far:
 - I agree with @keller-mark re: the use of `useId`, especially for components which may appear more than once on a page; does the current solution work if there's a second vitessce instance further down the page? 
 - I also second preferring to disable controls rather than conditionally render them. Controls popping in and out of existence is detrimental to user experience, especially for users of screen-readers.
 - I think a lot of the places where `key` is currently in use may be better served with an `id` instead.
 - Slider accessibility may be further enhanced in cases where there are multiple thumbs on one slider by having dynamic aria labels for each thumb.

NickAkhmetov reviewed (commented):
After reviewing the rest of the changes, the feedback so far should cover just about everything I saw, i.e.:
- Opt for `disabled` over conditional rendering
- Use `id` instead of `key` in most places where a key has been added
- Consider `useId` for anything that can appear more than once on a page

ivababukova commented on packages/view-types/heatmap/src/HeatmapOptions.js:
```diff
@@ -69,19 +70,23 @@ export default function HeatmapOptions(props) {
             value={geneExpressionColormapRange}
             onChange={handleColormapRangeChangeDebounced}
-            aria-labelledby="gene-expression-colormap-range-slider"
+            aria-labelledby="heatmap-gene-expression-colormap-range"
```
fixed in commit a4e69d5a

ivababukova commented on packages/view-types/layer-controller/src/ColorPalette.js:
```diff
@@ -33,6 +33,7 @@ const ColorPalette = ({ handleChange }) => {
           className={classes.button}
           key={color}
           onClick={() => handleChange(color)}
+          aria-label="Change colour"
```
we should be consistent. I have changed it to US convention

ivababukova commented on packages/view-types/scatterplot-gating/src/GatingScatterplotOptions.js:
```diff
@@ -107,19 +107,20 @@ export default function GatingScatterplotOptions(props) {
+        <TableCell className={classes.inputCell} variant="body">
           <TextField
-            label="Number"
+            label="Trasnform Coefficient"
```
fixed here: c6ee12fa

ivababukova commented on packages/view-types/scatterplot/src/shared-spatial-scatterplot/ToolMenu.js:
```diff
@@ -136,6 +136,7 @@ export default function ToolMenu(props) {
       <IconButton
         alt="click to recenter"
         onClick={() => onRecenterButtonCLick()}
+        aria-label="View recenter button"
```
Thanks for the suggestion. I applied it as part of 79194723 commit

ivababukova commented:
@keller-mark and @NickAkhmetov thank you both for the detailed look at this PR and the comments. I have addressed all of them. @NickAkhmetov in terms of your notes (https://github.com/vitessce/vitessce/pull/1612#pullrequestreview-1575856480 and https://github.com/vitessce/vitessce/pull/1612#pullrequestreview-1572846694):

1. The use of `useId()` React hook: I have went over the code and made sure that whatever component is called many times, it uses it. I found that this is the case for all the cases apart from Tooltip component. In most of the other places however, we use `uuidv4()`. I didn't refactor it, because it is still working as is and I didn't want to increase the scope of this PR even further. Let me know if you disagree though, I am happy to make the change.
2. I see that both of you think that using `disabled` is better than showing and hiding options, so I have brought this back (commits e5371cbf and e9a8ad07).
3. Use of `id` instead of `key` -- done
4. On the Slider accessibility -- I have gone over all the sliders, made sure that their labeling mechanism is consistent and added dynamic aria labels for each thumb in the cases of multiple thumbs. 

Could you take another look and let me know if there is anything else that needs fixing or if I should merge this in?

keller-mark commented on packages/tooltip/src/Tooltip.js:
```diff
@@ -42,6 +43,7 @@ export default function Tooltip(props) {
           container={getTooltipContainer}
           transition
           placement={`${placementY}-${placementX}`}
+          id={tooltipId}
```
I don't think adding IDs on their own affects the accessibility so I would revert this. 

The IDs are useful when other things reference them, such as `htmlFor` properties on label elements which reference the IDs of their corresponding input elements.

keller-mark commented on packages/tooltip/src/Tooltip2D.js:
```diff
@@ -37,6 +37,7 @@ export default function Tooltip2D(props) {
   } = props;
 
   const classes = useStyles();
+  const tooltipId = useId();
```
See above comment about IDs without corresponding references in other properties

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -35,62 +35,69 @@ export default function FeatureListOptions(props) {
           >
             {FEATURELIST_SORT_OPTIONS.map(option => (
-              <option key={option} value={option}>{option}</option>
+              <option id={option} value={option}>{option}</option>
```
```suggestion
              <option key={option} value={option}>{option}</option>
```

The key here is because we are using `.map` - please see https://react.dev/learn/rendering-lists#rules-of-keys

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -35,62 +35,69 @@ export default function FeatureListOptions(props) {
                   <>
-                    <option value="featureLabels">{primaryColumnName}</option>
-                    <option value="featureIndex">{ALT_COLNAME}</option>
+                    <option id={`featureLabels-${primaryColumnName}-key`} value="featureLabels">{primaryColumnName}</option>
+                    <option id={`featureIndex-${ALT_COLNAME}-key`} value="featureIndex">{ALT_COLNAME}</option>
```
Please revert, I dont think these IDs are doing anything

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -35,62 +35,69 @@ export default function FeatureListOptions(props) {
                 inputProps={{
-                  id: 'feature-list-sort-key-select',
+                  'aria-label': 'Select the feature list sort key',
+                  id: 'feature-list-sort-key',
```
this ID should come from `useId()`

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -35,62 +35,69 @@ export default function FeatureListOptions(props) {
-            <TableCell>
+            <TableCell variant="body">
               <OptionSelect
+                key="feature-list-sort-key-select"
```
Please remove

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -35,62 +35,69 @@ export default function FeatureListOptions(props) {
             onChange={handleFeatureListSortChange}
             inputProps={{
-              id: 'feature-list-sort-option-select',
+              id: 'feature-list-sort-option',
```
This ID should come from `useId()`

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -35,62 +35,69 @@ export default function FeatureListOptions(props) {
                   </>
                 ) : (
-                  <option value="featureIndex">{primaryColumnName}</option>
+                  <option id={`featureIndex-${primaryColumnName}-key`} value="featureIndex">{primaryColumnName}</option>
```
ditto

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -35,62 +35,69 @@ export default function FeatureListOptions(props) {
                 color="default"
+                inputProps={{
+                  'aria-label': 'Checkbox for showing or hiding alternative feature ids.',
+                  id: 'feature-list-show-alternative-ids',
```
should come from useId()

keller-mark commented on packages/view-types/heatmap/src/HeatmapOptions.js:
```diff
@@ -35,16 +35,17 @@ export default function HeatmapOptions(props) {
             inputProps={{
-              id: 'gene-expression-colormap-select',
+              'aria-label': 'Select gene expression colormap.',
+              id: 'heatmap-gene-expression-colormap',
```
should come from useId()

keller-mark commented on packages/view-types/heatmap/src/HeatmapOptions.js:
```diff
@@ -67,21 +68,25 @@ export default function HeatmapOptions(props) {
             color="default"
+            inputProps={{
+              'aria-label': 'Checkbox for showing or hiding tooltips.',
+              id: 'gene-expression-colormap-option-tooltip-visibility',
```
should come from useId()

keller-mark commented on packages/view-types/layer-controller/src/LayerController.js:
```diff
@@ -476,6 +476,8 @@ export default function LayerController(props) {
       TransitionProps={{ enter: false }}
       expanded={!disabled && isExpanded}
+      aria-controls="layer-controls options"
+      id="layer-controls-accordion"
```
should come from useId

keller-mark commented on packages/view-types/layer-controller/src/styles.js:
```diff
@@ -9,6 +9,16 @@ const useSpanStyles = makeStyles(() => ({
   },
 }));
 
+export const channelSliderStyles = makeStyles(theme => ({
```
```suggestion
export const useChannelSliderStyles = makeStyles(theme => ({
```

Prefer to keep the `use` prefix naming convention for makeStyles

keller-mark commented on packages/view-types/scatterplot-gating/src/GatingScatterplotOptions.js:
```diff
@@ -61,19 +61,19 @@ export default function GatingScatterplotOptions(props) {
             onChange={handleGeneSelectChange}
             inputProps={{
-              id: 'scatterplot-gene-select',
+              id: 'scatterplot-gating-gene-select',
```
should come from useId

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -92,24 +92,28 @@ export default function ScatterplotOptions(props) {
             color="default"
+            inputProps={{
+              'aria-label': 'Checkbox for showing/hiding set labels.',
+              id: 'scatterplot-set-labels-visible',
```
should come from useId

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -119,22 +123,26 @@ export default function ScatterplotOptions(props) {
             color="default"
+            inputProps={{
+              'aria-label': 'Checkbox for showing/hiding tooltips.',
+              id: 'scatterplot-set-tooltips-visible',
```
should come from useId

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -143,30 +151,35 @@ export default function ScatterplotOptions(props) {
             onChange={handleCellRadiusModeChange}
             inputProps={{
-              id: 'cell-radius-mode-select',
+              id: 'set-radius-mode-select',
```
should come from useId

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -35,62 +35,69 @@ export default function FeatureListOptions(props) {
-        <TableCell className={classes.labelCell} htmlFor="feature-list-sort-option-select">
+        <TableCell className={classes.labelCell} variant="head" scope="row">
           Sort Ordering
         </TableCell>
-        <TableCell>
+        <TableCell variant="body">
           <OptionSelect
+            key="feature-list-sort-option-select"
             className={classes.select}
             value={featureListSort}
             onChange={handleFeatureListSortChange}
             inputProps={{
-              id: 'feature-list-sort-option-select',
+              id: 'feature-list-sort-option',
```
Hi @ivababukova ,

In general, whenever there is a component with this structure:

```
<TableRow>
  <TableCell>
     Some Label Text
  </TableCell>
  <TableCell>
    <SomeInput />
  </TableCell>
</TableRow>
```
Please do the following (if the MUI component supports `inputProps` then use that, otherwise supply `id={someId}` as a direct prop):


```
import { useId } from 'react';

// ...
const someId = useId();
// ...
<TableRow>
  <TableCell variant="head" scope="row">
     <label htmlFor={someId}>Some Label Text</label>
  </TableCell>
  <TableCell variant="body">
    <SomeInput inputProps={{ id: someId }} />
  </TableCell>
</TableRow>
```

(I am omitting extra props like `className`s here but please keep those)

For background, please see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label#attributes

ivababukova replied:
It is very strange, because I was sure I actually did that. I don't know what happened but my changes must have got lost. I have fixed all of that now. I also had to update the eslint rules in relation to this.

NickAkhmetov commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -35,16 +37,17 @@ export default function FeatureListOptions(props) {
-        <TableCell className={classes.labelCell} htmlFor="feature-list-sort-option-select">
-          Sort Ordering
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label htmlFor={['feature-list-sort-option', featureListId].join('-')}>Sort Ordering</label>
```
I think it might be a bit clearer to use a template string than an array, e.g.:

```suggestion
          <label htmlFor={`feature-list-sort-option-${featureListId}`}>Sort Ordering</label>
```

NickAkhmetov commented on packages/view-types/layer-controller/src/ChannelOptions.js:
```diff
@@ -44,14 +44,25 @@ function ChannelOptions({ handlePropertyChange, handleChannelRemove, handleIQRUp
+      ariaLabel="Open channel options menu."
     >
-      <MenuItem dense disableGutters onClick={handleRemove}>
+      <MenuItem dense disableGutters onClick={handleRemove} aria-label="Click to remove handle channel">
```
Should this be `Click to remove channel`?

NickAkhmetov commented on packages/view-types/layer-controller/src/ImageAddButton.js:
```diff
@@ -37,6 +37,7 @@ function ImageAddButton({ imageOptions, handleImageAdd }) {
       buttonIcon={<ImageAddIcon />}
       buttonClassName={classes.addButton}
       placement="bottom-start"
+      ariaLabel="Add image menu."
```
Not a huge concern, but we should have consistency re: period usage in aria labels. I think we can probably drop the period since labels are meant to be short and concise; we can use `aria-description` for any longer descriptions that would require more than one sentence

NickAkhmetov commented on packages/view-types/layer-controller/src/LayerController.js:
```diff
@@ -476,6 +477,8 @@ export default function LayerController(props) {
       }
       TransitionProps={{ enter: false }}
       expanded={!disabled && isExpanded}
+      aria-controls="layer-controls options"
```
This attribute is supposed to point to the ID of whatever element a given HTML element controls; e.g. with tab panels, tabs get `aria-controls` attributes corresponding to whichever panel they're associated with. 

Looking at what the [MUI docs suggest](https://v4.mui.com/components/accordion/#accessibility), their recommendation is to put `aria-controls` on the `AccordionSummary` to indicate which panel the header controls. The `AccordionSummary` here already has this value set to `layer-${name}-controls` but I couldn't find any matching ID's for that value - maybe that should be provided to the `AccordionDetails`?

ivababukova replied:
I read this page a few times, the text is super brief, but the way I understand it is similar: that the `id` of `AccordionDetails` has to be the same as `aria-controls` of `AccordionSummary`. I have now modified the code to do that.

NickAkhmetov reviewed (commented):
Great job, thank you for making those corrections!

I have a few remaining concerns:
- I'm not sure that using `['string-1', 'string-2'].join('-')` is the best for clarity in defining ID's; could we use [template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) instead? 
- ARIA labels should be brief (same as any input label, ideally); if there's a full sentence/if a period is needed, using `aria-description` is a good way to provide any extra context while keeping the label itself short and to the point.

NickAkhmetov reviewed (approved):
Great work, thank you for making those revisions! I have no further concerns; if @keller-mark has no additional comments, this should be good to merge in my opinion! 👏

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -92,24 +94,36 @@ export default function ScatterplotOptions(props) {
-          {observationsLabelNice} Set Labels Visible
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-set-labels-visible', scatterplotOptionsId].join('-')}
```
```suggestion
            htmlFor={`scatterplot-set-labels-visible-${scatterplotOptionsId}`}
```

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -92,24 +94,36 @@ export default function ScatterplotOptions(props) {
             color="default"
+            inputProps={{
+              'aria-label': 'Show or hide set labels',
+              id: ['scatterplot-set-labels-visible', scatterplotOptionsId].join('-'),
```
```suggestion
              id: `scatterplot-set-labels-visible-${scatterplotOptionsId}`,
```

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -92,24 +94,36 @@ export default function ScatterplotOptions(props) {
-          Tooltips Visible
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-set-tooltips-visible', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -119,22 +133,31 @@ export default function ScatterplotOptions(props) {
             color="default"
+            inputProps={{
+              'aria-label': 'Show or hide tooltips',
+              id: ['scatterplot-set-tooltips-visible', scatterplotOptionsId].join('-'),
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -119,22 +133,31 @@ export default function ScatterplotOptions(props) {
-          {observationsLabelNice} Set Label Size
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-set-label-size', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -119,22 +133,31 @@ export default function ScatterplotOptions(props) {
             onChange={handleLabelSizeChange}
-            aria-labelledby="cell-set-label-size-slider"
+            getAriaLabel={() => 'Scatterplot label size slider'}
+            id={['scatterplot-set-label-size', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -143,30 +166,43 @@ export default function ScatterplotOptions(props) {
-          {observationsLabelNice} Set Polygons Visible
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-set-polygons-visible', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -143,30 +166,43 @@ export default function ScatterplotOptions(props) {
             color="default"
+            inputProps={{
+              'aria-label': 'Show or hide polygons',
+              id: ['scatterplot-set-polygons-visible', scatterplotOptionsId].join('-'),
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -143,30 +166,43 @@ export default function ScatterplotOptions(props) {
-          {observationsLabelNice} Radius Mode
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-set-radius-mode-select', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -143,30 +166,43 @@ export default function ScatterplotOptions(props) {
             onChange={handleCellRadiusModeChange}
             inputProps={{
-              id: 'cell-radius-mode-select',
+              id: ['scatterplot-set-radius-mode-select', scatterplotOptionsId].join('-'),
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -175,16 +211,21 @@ export default function ScatterplotOptions(props) {
-          {observationsLabelNice} Radius
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-set-radius-size-select', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -175,16 +211,21 @@ export default function ScatterplotOptions(props) {
             onChange={handleRadiusChange}
-            aria-labelledby="cell-radius-slider"
+            getAriaLabel={() => 'Scatterplot radius size slider'}
+            id={['scatterplot-set-radius-size-select', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -193,16 +234,21 @@ export default function ScatterplotOptions(props) {
-          {observationsLabelNice} Opacity Mode
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-set-opacity-mode', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -193,16 +234,21 @@ export default function ScatterplotOptions(props) {
             onChange={handleCellOpacityModeChange}
             inputProps={{
-              id: 'cell-opacity-mode-select',
+              id: 'scatterplot-set-opacity-mode',
```
Missing ID

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -211,16 +257,21 @@ export default function ScatterplotOptions(props) {
-          {observationsLabelNice} Opacity
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-set-opacity-level', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -211,16 +257,21 @@ export default function ScatterplotOptions(props) {
             onChange={handleOpacityChange}
-            aria-labelledby="cell-opacity-slider"
+            getAriaLabel={() => 'Scatterplot opacity level slider'}
+            id={['scatterplot-set-opacity-level', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -229,16 +280,21 @@ export default function ScatterplotOptions(props) {
-          Gene Expression Colormap
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-gene-expression-colormap', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -229,16 +280,21 @@ export default function ScatterplotOptions(props) {
             onChange={handleGeneExpressionColormapChange}
             inputProps={{
-              id: 'gene-expression-colormap-select',
+              id: 'scatterplot-gene-expression-colormap',
```
Missing ID

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -248,15 +304,23 @@ export default function ScatterplotOptions(props) {
-          Gene Expression Colormap Range
+        <TableCell className={classes.labelCell} variant="head" scope="row">
+          <label
+            htmlFor={['scatterplot-gene-expression-colormap-range', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on packages/view-types/scatterplot/src/ScatterplotOptions.js:
```diff
@@ -248,15 +304,23 @@ export default function ScatterplotOptions(props) {
+              const labelPrefix = index === 0 ? 'Low value slider' : 'High value slider';
+              return `${labelPrefix} for scatterplot gene expression colormap range`;
+            }}
+            id={['scatterplot-gene-expression-colormap-range', scatterplotOptionsId].join('-')}
```
Please update to other syntax

keller-mark commented on sites/docs/docs/tutorial-plugin-coordination-type.mdx:
```diff
@@ -40,6 +40,7 @@ function MyPluginSlider(props) {
         min={0.0}
         max={1.0}
         step={0.005}
+        aria-label="custom-coordination-type-slider"
```
```suggestion
        getAriaLabel={() => 'custom-coordination-type-slider'}
```

keller-mark commented on packages/view-types/feature-list/src/FeatureListOptions.js:
```diff
@@ -81,16 +85,20 @@ export default function FeatureListOptions(props) {
-            <TableCell className={classes.labelCell}>
-              Show Alternate IDs
+            <TableCell className={classes.labelCell} variant="head" scope="row">
+              <label htmlFor={`${'feature-list-show-alternative-ids'}-${featureListId}`}>Show Alternate IDs</label>
```
```suggestion
              <label htmlFor={`feature-list-show-alternative-ids-${featureListId}`}>Show Alternate IDs</label>
```
